[theia] update, unvendor Spectra, drop [viewer] from openimageio

valgur opened the pull request:
Related to #39297. Opening a separate PR to limit that PR's scope a bit.

- Updated to the latest head commit version since the project has been inactive for several years and a new version incorporating these changes is unlikely to be published.
- Unvendored ~~`vlfeat`~~ and `spectra`. Edit: reverted unvendoring of `vlfeat` because it caused segfaults in unit tests.
- Replaced `find_package()` with `find_dependency()` in `TheiaConfig.cmake.in`.
- Dropped `viewer` feature from `openimageio`. The library uses OpenImageIO only for image reading/writing. The viewer drags in a heavy Qt dependency. Reverts the change from #34699.
- Added `suitesparse` as an explicit dependency since it is being used directly, not just through `ceres`.

The vlfeat patch can be dropped if the fix in #39379 is merged.

----
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] SHA512s are updated for each updated download.
- [x] The "supports" clause reflects platforms that may be fixed by this new version.
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

valgur committed:
[theia] merge patches, use find_dependency(), drop [viewer] from openimageio

valgur committed:
[theia] update vdb

valgur committed:
[theia] test with relaxed "supports" configuration

valgur committed:
[theia] update vdb

valgur committed:
[theia] remove unused CMake vars

valgur committed:
[theia] upgrade to head version 2019-07-17, unvendor vlfeat

valgur committed:
[theia] update vdb

valgur committed:
[theia] add a patch to replace POSIX getline()

valgur committed:
[theia] vdb

valgur committed:
[theia] Windows x86 and arm64 are not supported

valgur committed:
[theia] number patch filenames

valgur committed:
[theia] unvendor Spectra

valgur committed:
vdb

dg0yt commented on ports/theia/003-vlfeat-includes.patch:
```diff
@@ -0,0 +1,63 @@
+ 
+ extern "C" {
+-#include <vl/sift.h>
++#include <vlfeat/sift.h>
```
Comparing with https://packages.debian.org/sid/amd64/libvlfeat-dev/filelist, my conclusion is that port `vlfeat` doesn't install the headers to an unusual location.

But I have no idea if vcpgk maintainers would accept a correction.

dg0yt reviewed (commented):
I don't want to go into details now, but it seems to me that the dependencies patch should be revised more thoroughly, to do only required and minimal changes.
And DLL concerns might go to another patch.

valgur commented:
Ok, no problem, I'll try to slim down the patch to a minimum.

valgur committed:
[vlfeat] minimize 001-fix-external-dependencies.patch

valgur committed:
[vlfeat] vdb

BillyONeal commented on ports/theia/003-vlfeat-includes.patch:
```diff
@@ -0,0 +1,63 @@
+Fix vlfeat include paths after unvendoring.
```
I think this patch should be reverted given that https://github.com/microsoft/vcpkg/pull/39379 was merged

valgur committed:
[theia] drop vlfeat includes patch

valgur committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into update/theia

valgur committed:
vdb

JonLiu1993 commented on versions/t-/theia.json:
```diff
@@ -1,5 +1,15 @@
+      "version-date": "2019-07-17",
+      "port-version": 0
+    },
+    {
+      "git-tree": "0b39f2dae77a0953ab31db028356998e650f02f2",
+      "version": "0.8",
+      "port-version": 11
+    },
```
```suggestion
```

JonLiu1993 replied:
We can only add one "git-tree" each time we merge

BillyONeal commented on ports/theia/005-replace-posix-getline.patch:
```diff
@@ -0,0 +1,63 @@
++std::string getline(FILE* file) {
++    std::string line;
++    char buffer[1024];
++    while (fgets(buffer, sizeof(buffer), file)) {
```
This implementation is very wrong unless all lines are exactly 1024 chars long. Did upstream see this?

valgur replied:
Are you sure? fgets reads a line or at most n characters, whichever comes first. The loop is there to handle lines of any length.

I have not submitted it upstream yet. I'll definitely open a PR, but I don't have high hopes since the project has been inactive for several years, unfortunately.

BillyONeal replied:
>Are you sure? fgets reads a line or at most n characters, whichever comes first.

Huh, I guess I forgot about that.

>I have not submitted it upstream yet. I'll definitely open a PR, but I don't have high hopes since the project has been inactive for several years, unfortunately.

We generally want to give upstream 30 days to respond to changes unless they're 'blindingly obvious' (like adding missing `#include`) or specific to how a thing is packaged with us (like fixing up CMake config or package names). We aren't the upstream maintainers and don't want them to dislike us because we're used as a way to 'go around' their desires with the things they own, if that makes sense.

valgur committed:
[theia] revert vlfeat unvendoring

Caused segfaults in the project's unit tests:
sift_descriptor_test
sift_detector_test

valgur committed:
vdb

valgur committed:
[theia] osx_arm64 is not supported

The vendored vlfeat fails to compile for some reason.

valgur committed:
vdb

Commit:
Merge branch 'master' of https://github.com/microsoft/vcpkg into update/theia

JonLiu1993 commented:
@valgur, Have you submitted a PR upstream?

valgur commented:
Theia has been unlisted from Vcpkg.
